### PR TITLE
Input can be a path prefix

### PR DESCRIPTION
**Pull request by nh13**
The user may provide a path prefix (or directory) in which undetermined FASTQs live.  The file names are assumed to be of the form: `<*>_L00#_<R# or I#>_001.fastq.gz`.  The following can be inferred from this pattern: the lane number, read number, and if the FASTQ is a template or index read.

If no read structure is given either on the command line or Sample Sheet, then a new list of read structure is built, one per input FASTQ, where the _kind_ (segment type) is inferred from the file name and all bases are used.  E.g.
```
test_L001_I1.fastq.gz
test_L001_R1.fastq.gz
test_L001_R2.fastq.gz
test_L001_I2.fastq.gz
```
would yield the read structures `+B +T +T +B`.  Reads will all molecular barcodes are supported (e.g. `test_L001_U1.fastq.gz`) and bases to skip (e.g. `test_L001_S1.fastq.gz`).  The utility of the latter is questionable, but kept for completeness.

Improvement: due to the need for all sample barcode read segments to have a _fixed length_ (for index hopping metrics collection and some validation), the first read is read from each FASTQ that contains a variable length sample barcode to determine the read length and fix the length of the variable length sample barcode.  I don't believe that a read structure like `+B +T +T +B` previously worked!

### CODE REVIEW

File: src/lib/opts.rs [hunk 1]
```diff
@@ -249,6 +249,65 @@ impl Opts {
         }
         Ok(output_types_to_write)
     }
+
```
Comment (nh13) [inline src/lib/opts.rs:252]: The following method is needed since index hopping metrics and some command line option validation requires the length of every sample barcode segment to be known.

File: src/lib/run.rs [hunk 2]
```diff
@@ -65,6 +83,15 @@ pub fn run(opts: Opts) -> Result<(), anyhow::Error> {
         samples.iter().all(|s| s.barcode.len() == samples[0].barcode.len()),
         "Sample metadata barcodes are unequal lengths."
     );
+
+    // If there is a read structure that's all sample barcode, we need to replace it with the
+    // expected length to enable index hopping metrics.  Do so by inspecting the first read in the
+    // corresponding FASTQ
+    // TODO: move to it's own method to test
```
Comment (nh13) [inline src/lib/run.rs:90]:
```suggestion
```

File: README.md [hunk 3]
```diff
@@ -126,6 +127,18 @@ for read in R1 R2 I1 I2; do cat L*/${read}.fastq.gz > ./${read}.fastq.gz; done
 
 FASTQ files _must_ be BGZF compressed.
 
+All reads containing sample barcodes/indexes _must_ be of the same length.
```
Comment (tfenne) [inline README.md:130]: I think this actually confuses things.  I know what you're trying to say here because of the context of this PR, but are you really saying that if I have a fastq with read structure `8B+T` that it's now required that all reads from the matching fastq must be a fixed length?

I think what you really mean is that sample barcode segments in read structures must be of fixed lengths.

File: README.md [hunk 4]
```diff
@@ -126,6 +127,18 @@ for read in R1 R2 I1 I2; do cat L*/${read}.fastq.gz > ./${read}.fastq.gz; done
...
+
+###### Auto-detecting FASTQS from a Path Prefix
+
+Alternatively, the FASTQS can be auto-detected when a path prefix is given to `--fastqs <dir>/<prefix>`.
+The FASTQs must be named `<dir>/<prefix>_L00<lane>_<kind><kind-number>_001.fastq.gz`, where `kind` is
+one of R (read/template), I (index/sample barcode), or U (umi/molecular barcode). 
+
+If the read structure is given on the command line or Sample sheet, the segments are assumed to to
+apply to I1 R1 R2 I2 reads in that order.  Otherwise, the read structure will be `B+ T+ T+ B+`
```
Comment (tfenne) [inline README.md:139]:
```suggestion
apply to I1 R1 R2 I2 reads in that order.  Otherwise, the read structure will be `+B +T +T +B` 
```

This feels very strange/dangerous to me.  Firstly I would have totally expected a different ordering (R1 I1 I2 R2), but moreso that if we're auto-detecting the FASTQs I don't think we should let the users specify the read-structures - it's just too hard to ensure they are in sync.  Either specify both, or auto both would be my preference.

File: src/lib/opts.rs [hunk 5]
```diff
@@ -79,7 +78,7 @@ For support please contact: care@singulargenomics.com
 #[derive(Parser, Debug, Clone)]
 #[clap(name = TOOL_NAME, version = built_info::VERSION.as_str(), about=SHORT_USAGE, long_about=LONG_USAGE, term_width=0)]
 pub struct Opts {
-    /// Path to the input FASTQs.
+    /// Path to the input FASTQs or path prefix.
```
Comment (tfenne) [inline src/lib/opts.rs:81]: Maybe mention how you tell the two apart?  Do you assume any path that's not an extant file is a prefix?  Or is it based on file names/extensions.

File: src/lib/opts.rs [hunk 6]
```diff
@@ -91,7 +90,7 @@ pub struct Opts {
     #[clap(long, short = 'r', display_order = 3, required = true, multiple_values = true)]
     pub read_structures: Vec<ReadStructure>,
 
-    /// The directory to write outputs, the directory must exist.
+    /// The directory to write outputs.
```
Comment (tfenne) [inline src/lib/opts.rs:93]:
```suggestion
    /// The directory to write outputs to.
```

File: src/lib/opts.rs [hunk 7]
```diff
@@ -250,6 +249,65 @@ impl Opts {
...
+                if read_structure
+                    .iter()
+                    .all(|s| s.kind != SegmentType::SampleBarcode || s.has_length())
+                {
+                    // No variable length, so just clone the current read structure
+                    read_structures.push(read_structure.clone());
+                } else {
+                    // Get the read length from the FASTQ, subtract all non variable length
+                    // segments (must be a sample barcode if we've gone this far).
+                    let read_length: usize = infer_fastq_sequence_length(fastq.clone())?;
```
Comment (tfenne) [inline src/lib/opts.rs:282]: Why not just pass `fastq.as_ref()` or `&fastq`?

File: src/lib/opts.rs [hunk 8]
```diff
@@ -250,6 +249,65 @@ impl Opts {
...
+                    let fixed_length: usize =
+                        read_structure.iter().map(|s| s.length().unwrap_or(0)).sum();
+
+                    let read_segments: Vec<ReadSegment> = read_structure
+                        .iter()
+                        .filter_map(|s| {
+                            if s.has_length() {
+                                Some(*s)
+                            } else if fixed_length == read_length {
+                                None
```
Comment (tfenne) [inline src/lib/opts.rs:292]: Do we really want to allow this?  This seems like it should only happen if there's an error somewhere.  Could check it on line 285 and return an Err instead?

File: src/lib/opts.rs [hunk 9]
```diff
@@ -295,3 +353,98 @@ pub fn setup() -> Opts {
...
+        // Create output FASTQ files
+        for fastq_def in &fastq_defs {
+            let writer = BufWriter::new(File::create(&fastq_def.fastq).unwrap());
+            let mut gz_writer = BgzfSyncWriter::new(writer, Compression::new(3));
+            let bases = String::from_utf8(vec![b'A'; fastq_def.read_length]).unwrap();
+            let quals = String::from_utf8(vec![b'I'; fastq_def.read_length]).unwrap();
+            let data = format!("@NAME\n{}\n+\n{}\n", bases, quals);
+            gz_writer.write_all(data.as_bytes()).unwrap();
+            gz_writer.flush().unwrap();
+            drop(gz_writer);
```
Comment (tfenne) [inline src/lib/opts.rs:430]: Why is this needed?  It should drop at the end of scope anyway.

File: src/lib/run.rs [hunk 10]
```diff
@@ -36,6 +35,25 @@ pub fn run(opts: Opts) -> Result<(), anyhow::Error> {
     // and so all accepted records will go into file(s) for one sample with no Undetermined
     let is_no_demux = samples.len() == 1 && samples[0].barcode.len() == 0;
 
+    // If a path prefix is given as the single argument to --fastqs then auto-detect all the FASTQs
+    // with that path prefix.  Also, if the read-structures is not given, build it from the auto-
+    // detected
+    let opts = if opts.fastqs.len() == 1 && !opts.fastqs[0].is_file() {
+        let input_fastqs: Vec<InputFastq> = InputFastq::slurp(opts.fastqs[0].clone());
```
Comment (tfenne) [inline src/lib/run.rs:42]: Pass a reference - no need to clone() paths.

File: src/lib/run.rs [hunk 11]
```diff
@@ -65,6 +83,14 @@ pub fn run(opts: Opts) -> Result<(), anyhow::Error> {
         samples.iter().all(|s| s.barcode.len() == samples[0].barcode.len()),
         "Sample metadata barcodes are unequal lengths."
     );
+
+    // If there is a read structure that's all sample barcode, we need to replace it with the
+    // expected length to enable index hopping metrics.  Do so by inspecting the first read in the
+    // corresponding FASTQ
+    let opts = opts.with_fixed_sample_barcodes()?;
```
Comment (tfenne) [inline src/lib/run.rs:90]: Maybe do this right after the auto-fastq block up on ~53?

File: src/lib/run.rs [hunk 12]
```diff
@@ -441,8 +476,33 @@ mod test {
...
+        match kind {
+            SegmentType::SampleBarcode => 'I',
+            SegmentType::Skip => 'S',
+            SegmentType::MolecularBarcode => 'U',
+            SegmentType::Template => 'R',
+            kind => panic!("Could not determine kind from {:?}", kind),
+        }
+    }
+
+    /// Creates the properly formatted FASZTQ file name to allow auto-detecting lane, kind, and
```
Comment (tfenne) [inline src/lib/run.rs:490]:
```suggestion
    /// Creates the properly formatted FASTQ file name to allow auto-detecting lane, kind, and
```

File: src/lib/utils.rs [hunk 13]
```diff
@@ -187,6 +198,187 @@ To compress an uncompressed FASTQ file with bgzip:
     Err(anyhow!(message).context(context))
 }
 
+pub static INPUT_FASTQ_SUFFIX: &str = "_001.fastq.gz";
+
+lazy_static! {
+    /// <*>_L00#_<R# or I#>_001.fastq.gz
+    static ref INPUT_FASTQ_REGEX: Regex = Regex::new(r"^(.*)_L00(\d{1})_([RIUS])(\d{1})_001.fastq.gz$").unwrap();
```
Comment (tfenne) [inline src/lib/utils.rs:205]: Isn't the `{1}` following `\d` redundant?

File: src/lib/utils.rs [hunk 14]
```diff
@@ -187,6 +198,187 @@ To compress an uncompressed FASTQ file with bgzip:
...
+    pub path: PathBuf,
+    pub prefix: String,
+    pub lane: usize,
+    pub kind: SegmentType,
+    pub kind_number: i32,
+}
+
+impl InputFastq {
+    /// Create a new `InputFastq` inferring information from the file name.  This must match the
+    /// `INPUT_FASTQ_REGEX` pattern.
```
Comment (tfenne) [inline src/lib/utils.rs:222]: says "must match", but the function returns `Option` and returns `Some` if the regex matches and `None` if it doesn't.

File: src/lib/utils.rs [hunk 15]
```diff
@@ -187,6 +198,187 @@ To compress an uncompressed FASTQ file with bgzip:
...
+    pub fn new<P: AsRef<Path>>(path: P) -> Option<InputFastq> {
+        let file_name = path.as_ref().file_name().unwrap().to_string_lossy();
+        match INPUT_FASTQ_REGEX.captures(&file_name) {
+            None => None,
+            Some(captures) => {
+                let prefix = captures.get(1).unwrap().as_str().to_string();
+                let lane = captures.get(2).unwrap().as_str().parse::<usize>().unwrap();
```
Comment (tfenne) [inline src/lib/utils.rs:229]: unwraps, unwraps everywhere!
Maybe add a comment as to why these unwrap()s are safe?

File: src/lib/utils.rs [hunk 16]
```diff
@@ -187,6 +198,187 @@ To compress an uncompressed FASTQ file with bgzip:
...
+                let kind: SegmentType = match captures.get(3).unwrap().as_str() {
+                    "R" => SegmentType::Template,
+                    "I" => SegmentType::SampleBarcode,
+                    "U" => SegmentType::MolecularBarcode,
+                    "S" => SegmentType::Skip,
+                    knd => panic!("Could not determine kind from {}", knd),
```
Comment (tfenne) [inline src/lib/utils.rs:235]: Given the regex, this should be impossible, right?  In which case maybe `panic!("Unreachable - regex should enforce only [RIUS]")`

File: src/lib/utils.rs [hunk 17]
```diff
@@ -187,6 +198,187 @@ To compress an uncompressed FASTQ file with bgzip:
...
+                };
+                let kind_number = captures.get(4).unwrap().as_str().parse::<i32>().unwrap();
```
Comment (tfenne) [inline src/lib/utils.rs:237]: Why i32?  Surely we shouldn't have negative read numbers?  So perhaps `u32`?

File: src/lib/utils.rs [hunk 18]
```diff
@@ -187,6 +198,187 @@ To compress an uncompressed FASTQ file with bgzip:
...
+                    kind_number,
+                })
+            }
+        }
+    }
+
+    /// Identifies all FASTQs that share the common path prefix and match the
+    /// `INPUT_FASTQ_REGEX` pattern.  The path prefix may also be a directory.  The FASTQS are
+    /// returned in sorted order.
+    pub fn slurp<P: AsRef<Path>>(path_prefix: P) -> Vec<InputFastq> {
```
Comment (tfenne) [inline src/lib/utils.rs:253]: perhaps `fn glob`?

File: src/lib/utils.rs [hunk 19]
```diff
@@ -187,6 +198,187 @@ To compress an uncompressed FASTQ file with bgzip:
...
+        let (parent, prefix) = if path_prefix.as_ref().is_file() {
+            (
+                path_prefix.as_ref().parent().unwrap(),
+                path_prefix.as_ref().file_name().unwrap().to_str().unwrap(),
```
Comment (tfenne) [inline src/lib/utils.rs:257]: This doesn't make a lot of sense to me.  `is_file()` should only return true for an extant file - I think this function only works if you pass in a directory with no other prefix as a result.

Also, if you inverted this (using `is_dir()`) these unwraps can very much fail.  E.g. with `--fastqs Undetermined` while the _file_ has a parent, the _path_ doesn't.

File: src/lib/utils.rs [hunk 20]
```diff
@@ -187,6 +198,187 @@ To compress an uncompressed FASTQ file with bgzip:
...
+            )
+        } else {
+            (path_prefix.as_ref(), "")
+        };
+
+        let mut fastqs: Vec<InputFastq> = std::fs::read_dir(parent)
+            .unwrap()
```
Comment (tfenne) [inline src/lib/utils.rs:264]: Perhaps you should be returning `Result<Vec<>>` and using `?`?

File: src/lib/utils.rs [hunk 21]
```diff
@@ -187,6 +198,187 @@ To compress an uncompressed FASTQ file with bgzip:
...
+}
+
+impl Ord for InputFastq {
+    /// Defines a partial ordering based on:
+    /// 1. The FASTQ prefix
+    /// 2. The lane
+    /// 3. The read number (kind number)
+    /// 4. The kind.  If the kind number is odd, orders by sample barcode, skip, molecular barcode,
+    ///    then template.  The opposite order if the kind is even.  This is so if the four FASTQS
+    ///    for I1, R1, R2, I2, they are remain in that order after sorting.
```
Comment (tfenne) [inline src/lib/utils.rs:304]: This is truly horrifying.

File: src/lib/utils.rs [hunk 22]
```diff
@@ -187,6 +198,187 @@ To compress an uncompressed FASTQ file with bgzip:
...
+impl PartialEq for InputFastq {
+    fn eq(&self, other: &Self) -> bool {
+        self.cmp(other).is_eq()
+    }
+}
+
+impl Eq for InputFastq {}
+
+/// Infers the read length contained in the given FASTQ by examining the length of the first read.
+pub fn infer_fastq_sequence_length(file: PathBuf) -> Result<usize, anyhow::Error> {
```
Comment (tfenne) [inline src/lib/utils.rs:354]: Just use anyhow's result type so you can say `Result<usize>`?

File: src/lib/utils.rs [hunk 23]
```diff
@@ -187,6 +198,187 @@ To compress an uncompressed FASTQ file with bgzip:
...
+            .num_threads(1)
+            .with_context(|| {
+                format!(
+                    "Error in setting threads when creating decompressor for {}",
+                    file.to_string_lossy()
+                )
+            })?
+            .from_reader(reader),
+        BUFSIZE,
+    );
```
Comment (tfenne) [inline src/lib/utils.rs:371]: Is there no simple non-par reader to use?

File: src/lib/utils.rs [hunk 24]
```diff
@@ -187,6 +198,187 @@ To compress an uncompressed FASTQ file with bgzip:
...
+
+    match reader.next() {
+        Some(Ok(record)) => Ok(record.seq().len()),
+        _ => Err(anyhow!(
+            "Could determine sample barcode length from input FASTQ: {}",
```
Comment (tfenne) [inline src/lib/utils.rs:376]:
```suggestion
            "Could determine sample barcode length from empty input FASTQ: {}",
```

**Review by tfenne — CHANGES_REQUESTED** [review]

File: README.md (hunk 3 above)
Comment (nh13) [reply README.md:130]: No, I really do mean this, I think.  To infer from the input FASTQ the read length so I can make variable length sample barcode segments of fixed lengths, I am going to look at the first read in the FASTQ (not all), so. want the read length in the FASTQ to be same length.

Comment (nh13) [reply README.md:130]: I am going to enforce that sample barcode segments must be fixed lengths

File: src/lib/utils.rs (hunk 19 above)
Comment (nh13) [reply src/lib/utils.rs:257]: Take one more look

File: src/lib/utils.rs (hunk 20 above)
Comment (nh13) [reply src/lib/utils.rs:264]: How does one avoid `unwrap` inside the `map/filter` closures?

File: src/lib/utils.rs (hunk 21 above)
Comment (nh13) [reply src/lib/utils.rs:304]: removed

File: Cargo.toml [hunk 25]
```diff
@@ -45,9 +45,11 @@ log = "0.4.14"
 mimalloc = { version = "0.1.27", default-features = false }
 num_cpus = "1.13.0"
 parking_lot = "0.11.2"
+path-absolutize = "3.0.14"
```
Comment (tfenne) [inline Cargo.toml:48]: It is _so_ rust that there's a crate for this and not use `Path.absolute()`

File: src/lib/opts.rs [hunk 26]
```diff
@@ -79,19 +78,19 @@ For support please contact: care@singulargenomics.com
...
     #[clap(long, short = 'f', display_order = 1, required = true, multiple_values = true)]
     pub fastqs: Vec<PathBuf>,
 
     /// Path to the sample metadata.
     #[structopt(long, short = 's', display_order = 2)]
     pub sample_metadata: PathBuf,
 
-    /// Read structures, one per input FASTQ.
+    /// Read structures, one per input FASTQ. Do not provide when using a path prefix for FASTQs.
     #[clap(long, short = 'r', display_order = 3, required = true, multiple_values = true)]
```
Comment (tfenne) [inline src/lib/opts.rs:90]: Still has required=true.  Probably need to drop that and separately validate that it's given _if_ a prefix is given?

File: src/lib/opts.rs (hunk 8 above)
Comment (tfenne) [reply src/lib/opts.rs:292]: I get that `+B` can technically resolve to 0 bases on any given read, but in this particular case wouldn't it be weird to have a read structure that has a `+B` in it and then have that resolve to zero bases?  And then omitting that segment from the read structure will cause super-confusing error messages because it will shift any later `B` segments down by one when comparing to the sample sheet/metadata.

Hence the suggestion to just error out earlier if the fixed-length segments add up to the read length, and we have a disappearing barcode segment.  I'd say you could insert a `0B` here but I see the read-structure crate disallows that (somewhat sensibly).

File: src/lib/run.rs [hunk 27]
```diff
@@ -36,6 +35,49 @@ pub fn run(opts: Opts) -> Result<(), anyhow::Error> {
...
+        // Important: sort by kind then kind number so the output kind number is ordered correctly
+        let input_fastqs: Vec<InputFastq> = opts
+            .fastqs
+            .iter()
+            .flat_map(|prefix| InputFastq::glob(&prefix).unwrap())
+            .sorted_by(|left, right| {
+                left.kind.cmp(&right.kind).then(left.kind_number.cmp(&right.kind_number))
+            })
+            .collect();
+
```
Comment (tfenne) [inline src/lib/run.rs:59]: Until we add multi-lane support, do you want to validate here that you only got 1 each of any read/kind combo?  I.e. you didn't get 2+ `R1` files?

File: src/lib/run.rs [hunk 28]
```diff
@@ -36,6 +35,49 @@ pub fn run(opts: Opts) -> Result<(), anyhow::Error> {
...
+        let read_structures: Vec<ReadStructure> =
+            input_fastqs.iter().map(InputFastq::read_structure).collect();
+
+        // extract the list of input FASTQs
+        let fastqs: Vec<PathBuf> =
+            input_fastqs.iter().map(|input_fastq| input_fastq.path.clone()).collect();
+
+        Opts { fastqs, read_structures, ..opts }
+    } else {
+        bail!("Input FASTQS (--fastqs) must either all be files or all path prefixes, not a mixture of both")
```
Comment (tfenne) [inline src/lib/run.rs:70]: Is it intentional that this code can work with _multiple_ path prefixes?  I don't think it's an issue, but it might be simpler to validate we only get one if it's a prefix?

File: src/lib/run.rs [hunk 29]
```diff
@@ -65,6 +107,15 @@ pub fn run(opts: Opts) -> Result<(), anyhow::Error> {
...
         "Sample metadata barcodes are unequal lengths."
     );
+
+    // All sample barcode read segments should now have a fixed length, so check the sum of their
+    // lengths with the sum of length of the sample barcode(s) in the sample sheet.
+    ensure!(
+        opts.read_structures
+            .iter()
+            .all(|s| s.sample_barcodes().all(read_structure::ReadSegment::has_length)),
+        "The Read Structure must have sample barcode segments with fixed lengths"
```
Comment (tfenne) [inline src/lib/run.rs:117]:
```suggestion
        "Sample barcode segments in read structures must have fixed lengths."
```

**Review by tfenne — APPROVED** [review]
A few minor comments and suggestions, but good to merge once you're happy.

File: Cargo.toml (hunk 25 above)
Comment (nh13) [reply Cargo.toml:48]: There's `is_absolute` on `Path`, but not `make absolute`.  Sigh

**Review by omicsorama — APPROVED** [review]
